logstash-core/lib/logstash/pipeline_reporter.rb

```diff
@@ -0,0 +1,119 @@
+require 'ostruct'
+
+# encoding: utf-8
+module LogStash; class PipelineReporter
+  attr_reader :logger, :pipeline
+
+  # This is an immutable copy of the pipeline state,
+  # It is a proxy to a hash to allow us to add methods dynamically to the hash
+  class Snapshot
+    def initialize(data)
+      @data = data
+    end
+
+    def to_hash
+      @data
+    end
+
+    def to_simple_hash
+      {"inflight_count" => inflight_count, "stalling_thread_info" => format_threads_by_plugin}
+    end
+
+    def method_missing(meth)
```

Comment:
Having read through these comments I agree with the concerns raised in the general case, but I don't think they apply in this specific case (as @guyboertje mentions).

@ph I like forwardable and def_delegator but I don't see what concrete benefit it provides here.

These methods will be called relatively infrequently, so the perf impact is negligible. I also don't see why any errors here would be hard to debug, because everything gets forwarded to a hash lookup. This is basically a decorated OStruct.